<a href="https://colab.research.google.com/github/data602sps/assignments/blob/master/05_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Assignment 7**

# **Weeks 8 & 9 - Pandas**
* In this homework assignment, you will explore and analyze a public dataset of your choosing. Since this assignment is “open-ended” in nature, you are free to expand upon the requirements below. However, you must meet the minimum requirments as indicated in each section. 

* You must use Pandas as the **primary tool** to process your data.

* The preferred method for this analysis is in a .ipynb file. Feel free to use whichever platform of your choosing.  
 * https://www.youtube.com/watch?v=inN8seMm7UI (Getting started with Colab).

* Your data should need some "work", or be considered "dirty".  You must show your skills in data cleaning/wrangling.

### **Some data examples:**
•	https://www.data.gov/

•	https://opendata.cityofnewyork.us/

•	https://datasetsearch.research.google.com/

•	https://archive.ics.uci.edu/ml/index.php

### **Resources:**

•	https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html 

•	https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html


### **Headings or comments**
**You are required to make use of comments, or headings for each section.  You must explain what your code is doing, and the results of running your code.**  Act as if you were giving this assignment to your manager - you must include clear and descriptive information for each section.

### **You may work as a group or indivdually on this assignment.**


# Introduction

For this assignment, I will be working with Income Data from the IRS for 2019. I will be using this data for my final project. 

Source: [US Dept of Treasury](https://catalog.data.gov/dataset/zip-code-data)

This is a large file (> 150mb), so I will be transforming the file in order to write a .csv which will meet GitHub filesize requirements (< 25mb). For this assignment, I will be reading the data locally from my machine.

______________
# Data Exploration
Import your dataset into your .ipynb, create dataframes, and explore your data.  

Include: 

* Summary statistics means, medians, quartiles, 
* Missing value information
* Any other relevant information about the dataset.  



In [35]:
import pandas as pd
import numpy as np

# file url
file = "/Users/joshiden/Documents/Classes/CUNY SPS/Fall 2022/DATA 602/DATA-602/FINAL_PROJECT/data.csv"
# read file into pandas dataframe
df = pd.read_csv(file)

In [25]:
# preview the dataframe
df.head()

Unnamed: 0  STATEFIPS STATE  zipcode  agi_stub        N1     mars1  \
0           0          1    AL        0         1  778210.0  491030.0   
1           1          1    AL        0         2  525940.0  247140.0   
2           2          1    AL        0         3  285700.0  105140.0   
3           3          1    AL        0         4  179070.0   38820.0   
4           4          1    AL        0         5  257010.0   28180.0   

      MARS2     MARS4       ELF  ...  N85300  A85300    N11901    A11901  \
0   84770.0  189600.0  712890.0  ...     0.0     0.0   62720.0   51936.0   
1  123910.0  139860.0  481760.0  ...     0.0     0.0   85860.0  122569.0   
2  128140.0   44560.0  260570.0  ...     0.0     0.0   73980.0  154932.0   
3  123110.0   13740.0  164300.0  ...     0.0     0.0   51330.0  139065.0   
4  216740.0    7150.0  236850.0  ...    90.0   141.0  104290.0  460071.0   

     N11900     A11900    N11902     A11902  N12000   A12000  
0  671860.0  1700965.0  669570.0  1694792.0  1980.0   3512.0  
1  438020.0  1274802.0  435210.0  1266557.0  3670.0   7410.0  
2  212040.0   575315.0  208470.0   564202.0  5020.0  13653.0  
3  126850.0   401581.0  123310.0   388749.0  3040.0  10377.0  
4  152790.0   598248.0  144640.0   539385.0  9180.0  56257.0  

[5 rows x 153 columns]

In [26]:
df.describe()

Unnamed: 0      STATEFIPS        zipcode       agi_stub  \
count  166159.000000  166159.000000  166159.000000  166159.000000   
mean    83079.000000      29.666885   48859.485553       3.499949   
std     47966.116026      15.121486   27167.679271       1.707871   
min         0.000000       1.000000       0.000000       1.000000   
25%     41539.500000      18.000000   27020.000000       2.000000   
50%     83079.000000      29.000000   48843.000000       3.000000   
75%    124618.500000      42.000000   70652.500000       5.000000   
max    166158.000000      56.000000   99999.000000       6.000000   

                 N1         mars1         MARS2         MARS4           ELF  \
count  1.661590e+05  1.661590e+05  1.661590e+05  166159.00000  1.661590e+05   
mean   1.860508e+03  9.127834e+02  6.478571e+02     258.03676  1.689549e+03   
std    3.722335e+04  2.224999e+04  1.200080e+04    6336.06430  3.347049e+04   
min    0.000000e+00  0.000000e+00  0.000000e+00       0.00000  0.000000e+00   
25%    7.000000e+01  0.000000e+00  4.000000e+01       0.00000  7.000000e+01   
50%    2.600000e+02  8.000000e+01  1.100000e+02      30.00000  2.400000e+02   
75%    1.080000e+03  3.900000e+02  3.800000e+02     100.00000  9.900000e+02   
max    5.506120e+06  4.069770e+06  1.818210e+06  945490.00000  4.827070e+06   

               CPREP  ...         N85300        A85300        N11901  \
count  166159.000000  ...  166159.000000  1.661590e+05  1.661590e+05   
mean       75.771821  ...      62.932011  3.016957e+02  3.982022e+02   
std      1866.726495  ...    3210.746066  1.727929e+04  7.639843e+03   
min         0.000000  ...       0.000000  0.000000e+00  0.000000e+00   
25%         0.000000  ...       0.000000  0.000000e+00  2.000000e+01   
50%         0.000000  ...       0.000000  0.000000e+00  6.000000e+01   
75%        40.000000  ...       0.000000  0.000000e+00  2.400000e+02   
max    338290.000000  ...  932390.000000  4.668052e+06  1.160480e+06   

             A11901        N11900        A11900        N11902        A11902  \
count  1.661590e+05  1.661590e+05  1.661590e+05  1.661590e+05  1.661590e+05   
mean   2.277891e+03  1.411865e+03  4.982449e+03  1.375670e+03  3.915695e+03   
std    7.750517e+04  2.939822e+04  1.051993e+05  2.903642e+04  7.570270e+04   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    2.200000e+01  5.000000e+01  1.680000e+02  5.000000e+01  1.510000e+02   
50%    1.480000e+02  1.900000e+02  6.470000e+02  1.800000e+02  5.720000e+02   
75%    6.690000e+02  7.700000e+02  2.523000e+03  7.400000e+02  2.238000e+03   
max    2.154141e+07  4.297720e+06  1.845451e+07  4.268070e+06  9.218146e+06   

              N12000        A12000  
count  166159.000000  1.661590e+05  
mean       43.742861  1.012980e+03  
std       991.034426  5.136345e+04  
min         0.000000  0.000000e+00  
25%         0.000000  0.000000e+00  
50%         0.000000  0.000000e+00  
75%        30.000000  6.200000e+01  
max    203100.000000  1.262769e+07  

[8 rows x 152 columns]

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 166159 entries, 0 to 166158
Columns: 153 entries, Unnamed: 0 to A12000
dtypes: float64(148), int64(4), object(1)
memory usage: 194.0+ MB


In [28]:
df.isna().sum()

Unnamed: 0    0
STATEFIPS     0
STATE         0
zipcode       0
agi_stub      0
             ..
A11900        0
N11902        0
A11902        0
N12000        0
A12000        0
Length: 153, dtype: int64

The dataset contains 166,158 observations (rows) of 153 variables (columns). For this assignment, I am going to focus my attention on five (5) columns. 

They are:

STATE - State  
ZIPCODE - Zip Code  
N1 - Number of Returns  
A00100 - Adjusted Gross Income  
A02650 - Total Income 

Subsetting and tidying this data will allow me to prepare better preliminary EDA.

# Data Wrangling
Create a subset of your original data and perform the following.  

1. Modify multiple column names.

2. Look at the structure of your data – are any variables improperly coded? Such as strings or characters? Convert to correct structure if needed.

3. Fix missing and invalid values in data.

4. Create new columns based on existing columns or calculations.

5. Drop column(s) from your dataset.

6. Drop a row(s) from your dataset.

7. Sort your data based on multiple variables. 

8. Filter your data based on some condition. 

9. Convert all the string values to upper or lower cases in one column.

10. Check whether numeric values are present in a given column of your dataframe.

11. Group your dataset by one column, and get the mean, min, and max values by group. 
  * Groupby()
  * agg() or .apply()

12. Group your dataset by two columns and then sort the aggregated results within the groups. 

**You are free (and should) to add on to these questions.  Please clearly indicate in your assignment your answers to these questions.**

In [77]:
# Subsetting the data to the desired columns
cols = ['STATE','zipcode','N1','A00100','A02650']
df_clean = df[cols]
df_clean.head()

STATE  zipcode        N1      A00100      A02650
0    AL        0  778210.0  10158838.0  10311099.0
1    AL        0  525940.0  18974967.0  19145621.0
2    AL        0  285700.0  17535801.0  17690402.0
3    AL        0  179070.0  15546951.0  15670456.0
4    AL        0  257010.0  34974856.0  35286228.0

In [96]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 166159 entries, 0 to 166158
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   STATE    166159 non-null  object 
 1   zipcode  166159 non-null  int64  
 2   N1       166159 non-null  float64
 3   A00100   166159 non-null  float64
 4   A02650   166159 non-null  float64
dtypes: float64(3), int64(1), object(1)
memory usage: 6.3+ MB


In [97]:
# rename columns
new_cols = {'STATE':'st','zipcode':'zip','N1':'n_returns','A00100':'agi','A02650':'total_income'}
df_clean.rename(new_cols, axis=1, inplace=True)
df_clean.head()

/var/folders/82/s2jpfjnx23x79yzgnlx3kx_h0000gp/T/ipykernel_61178/2504631388.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean.rename(new_cols, axis=1, inplace=True)


st  zip  n_returns         agi  total_income
0  AL    0   778210.0  10158838.0    10311099.0
1  AL    0   525940.0  18974967.0    19145621.0
2  AL    0   285700.0  17535801.0    17690402.0
3  AL    0   179070.0  15546951.0    15670456.0
4  AL    0   257010.0  34974856.0    35286228.0

It appears there are 0 values in the ZIPCODE column. Let's see how many of those there are.

In [98]:
# count total number of rows where ZIPCODE is 0
len(df_clean[df_clean['zip'] == 0])

306

Let's replace these zeros with NaN

In [99]:
# replace zeros with NaN
df_clean.zip.replace(0, np.nan, inplace=True)
df_clean.isna().sum()

/var/folders/82/s2jpfjnx23x79yzgnlx3kx_h0000gp/T/ipykernel_61178/3418041925.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean.zip.replace(0, np.nan, inplace=True)


st                0
zip             306
n_returns         0
agi               0
total_income      0
dtype: int64

Let's take a look at the top 5 observations of a few of the variables.

By Number of Returns

In [100]:
# sort by n_returns
df_clean.sort_values('n_returns', ascending=False).head()

st  zip  n_returns          agi  total_income
8470    CA  NaN  5506120.0   70433376.0    72029078.0
138160  TX  NaN  4572230.0   58343163.0    59509373.0
8471    CA  NaN  4401600.0  159743701.0   161747072.0
21784   FL  NaN  3837770.0   49642009.0    50737192.0
138161  TX  NaN  3203960.0  116205440.0   117492854.0

Odd that the areas with the largest number of returns do not specify ZIPCODE. 

By AGI

In [101]:
# sort by agi
df_clean.sort_values('agi', ascending=False).head()

st  zip  n_returns          agi  total_income
8475    CA  NaN  1477550.0  768358183.0   777314548.0
95512   NY  NaN   650530.0  400948012.0   406001116.0
8474    CA  NaN  2757790.0  382178470.0   386535464.0
138165  TX  NaN   736350.0  379735048.0   383222728.0
21789   FL  NaN   528650.0  341739166.0   344881211.0

By Total Income

In [102]:
# sort by total_income
df_clean.sort_values('total_income', ascending=False).head()

st  zip  n_returns          agi  total_income
8475    CA  NaN  1477550.0  768358183.0   777314548.0
95512   NY  NaN   650530.0  400948012.0   406001116.0
8474    CA  NaN  2757790.0  382178470.0   386535464.0
138165  TX  NaN   736350.0  379735048.0   383222728.0
21789   FL  NaN   528650.0  341739166.0   344881211.0

In [103]:
# calculate diff between total income and agi
df_clean['balance'] = df_clean.total_income - df_clean.agi
df_clean.head()

/var/folders/82/s2jpfjnx23x79yzgnlx3kx_h0000gp/T/ipykernel_61178/2789228845.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['balance'] = df_clean.total_income - df_clean.agi


st  zip  n_returns         agi  total_income   balance
0  AL  NaN   778210.0  10158838.0    10311099.0  152261.0
1  AL  NaN   525940.0  18974967.0    19145621.0  170654.0
2  AL  NaN   285700.0  17535801.0    17690402.0  154601.0
3  AL  NaN   179070.0  15546951.0    15670456.0  123505.0
4  AL  NaN   257010.0  34974856.0    35286228.0  311372.0

In [104]:
# convert all states to lower case
df_clean.st = df_clean.st.str.lower()
df_clean.head()

/var/folders/82/s2jpfjnx23x79yzgnlx3kx_h0000gp/T/ipykernel_61178/1742823452.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean.st = df_clean.st.str.lower()


st  zip  n_returns         agi  total_income   balance
0  al  NaN   778210.0  10158838.0    10311099.0  152261.0
1  al  NaN   525940.0  18974967.0    19145621.0  170654.0
2  al  NaN   285700.0  17535801.0    17690402.0  154601.0
3  al  NaN   179070.0  15546951.0    15670456.0  123505.0
4  al  NaN   257010.0  34974856.0    35286228.0  311372.0

In [105]:
# convert them back!
df_clean.st = df_clean.st.str.upper()
df_clean.head()

/var/folders/82/s2jpfjnx23x79yzgnlx3kx_h0000gp/T/ipykernel_61178/3266998985.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean.st = df_clean.st.str.upper()


st  zip  n_returns         agi  total_income   balance
0  AL  NaN   778210.0  10158838.0    10311099.0  152261.0
1  AL  NaN   525940.0  18974967.0    19145621.0  170654.0
2  AL  NaN   285700.0  17535801.0    17690402.0  154601.0
3  AL  NaN   179070.0  15546951.0    15670456.0  123505.0
4  AL  NaN   257010.0  34974856.0    35286228.0  311372.0

In [108]:
# group by st
df_clean.groupby('st')['n_returns'].mean().sort_values(ascending=False)

st
DC    4817.986111
CA    4135.121786
HI    3892.361111
NV    3881.602564
FL    3775.155628
AZ    3668.275666
DE    2807.807018
NJ    2717.736137
TX    2703.738539
RI    2560.117371
UT    2478.263158
MD    2472.204433
WA    2470.233333
MA    2385.325035
CO    2381.595118
GA    2339.987506
CT    2236.938131
NC    2178.614180
AK    2094.255952
NY    2093.093377
SC    2073.450736
OH    1902.753841
OR    1892.900188
MI    1790.177703
TN    1754.138418
VA    1694.217300
IL    1651.656060
IN    1567.448225
PA    1531.881248
NM    1500.993641
LA    1496.636971
WI    1343.136385
ID    1248.055556
AL    1216.500724
MN    1184.016013
MS    1136.135135
MO    1061.608785
NH    1032.893218
OK    1023.073394
KY     992.032770
AR     868.593964
WY     841.452599
MT     781.932075
KS     774.747126
NE     641.076171
ME     601.941441
IA     594.386391
SD     506.482800
WV     504.600524
VT     458.473980
ND     451.012346
Name: n_returns, dtype: float64

In [109]:
# get the min
# group by st
df_clean.groupby('st')['n_returns'].min().sort_values(ascending=False)

st
DC    40.0
AK     0.0
PA     0.0
ND     0.0
NE     0.0
NH     0.0
NJ     0.0
NM     0.0
NV     0.0
NY     0.0
OH     0.0
OK     0.0
OR     0.0
RI     0.0
MT     0.0
SC     0.0
SD     0.0
TN     0.0
TX     0.0
UT     0.0
VA     0.0
VT     0.0
WA     0.0
WI     0.0
WV     0.0
NC     0.0
MS     0.0
AL     0.0
MO     0.0
AR     0.0
AZ     0.0
CA     0.0
CO     0.0
CT     0.0
DE     0.0
FL     0.0
GA     0.0
HI     0.0
IA     0.0
ID     0.0
IL     0.0
IN     0.0
KS     0.0
KY     0.0
LA     0.0
MA     0.0
MD     0.0
ME     0.0
MI     0.0
MN     0.0
WY     0.0
Name: n_returns, dtype: float64

In [110]:
# group by st
df_clean.groupby('st')['n_returns'].max().sort_values(ascending=False)

st
CA    5506120.0
TX    4572230.0
FL    3837770.0
NY    3036960.0
PA    2019420.0
IL    1917000.0
OH    1867590.0
GA    1721640.0
MI    1680630.0
NC    1635030.0
NJ    1304360.0
VA    1182810.0
TN    1090830.0
IN    1088770.0
AZ    1036690.0
MO     965810.0
MA     944250.0
WA     931500.0
WI     867990.0
MD     860790.0
SC     826950.0
MN     784030.0
CO     781760.0
LA     779130.0
AL     778210.0
KY     693920.0
OR     605440.0
OK     581400.0
MS     519700.0
CT     512160.0
NV     491860.0
AR     477680.0
KS     428990.0
IA     428900.0
UT     427800.0
NM     355750.0
WV     286530.0
NE     277170.0
ID     260810.0
ME     209540.0
HI     199600.0
NH     190350.0
MT     175980.0
RI     172890.0
DE     147020.0
SD     134520.0
AK     102820.0
VT      99920.0
ND      98930.0
DC      82550.0
WY      81350.0
Name: n_returns, dtype: float64

In [111]:
# by two variables
df_clean.groupby(['st','zip'])['total_income'].max()

st  zip    
AK  99501.0    275002.0000
    99502.0    374708.0000
    99503.0    126531.0000
    99504.0    393118.0000
    99505.0     34534.0000
                  ...     
WY  83126.0      7132.0001
    83127.0     52911.0000
    83128.0    128067.0000
    83414.0     75902.0000
    99999.0    557343.0000
Name: total_income, Length: 27644, dtype: float64

# Conclusions  

After exploring your dataset, provide a short summary of what you noticed from this dataset.  What would you explore further with more time?

This data provides a birds eye view of income across states - it needs some more tidying. I need to understand how the total income and agi numbers related to the number of returns, as the average income does not compute when I try to calculate it. 